pokemon-service/src/main/java/cz/muni/fi/pa165/pokemon/league/participation/manager/service/TrainerServiceImpl.java

```diff
@@ -0,0 +1,171 @@
+package cz.muni.fi.pa165.pokemon.league.participation.manager.service;
+
+import cz.muni.fi.pa165.pokemon.league.participation.manager.dao.GymDAO;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.dao.TrainerDAO;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.entities.Pokemon;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.entities.Trainer;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.exceptions.NoAdministratorException;
+import java.math.BigInteger;
+import java.security.SecureRandom;
+import java.util.List;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import javax.inject.Inject;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service interface implementation for object Trainer.
+ *
+ * @author Jiří Medveď 38451
+ */
+@Service
+public class TrainerServiceImpl implements TrainerService {
+
+    @Inject
+    private TrainerDAO trainerDao;
+
+    @Inject
+    private GymDAO gymDao;
+
+    @Override
+    public boolean changePassword(Trainer trainer, String oldPassword, String newPassword) {
+        if (!authenticate(trainer, oldPassword)) {
+            return false;
+        }
+        trainer.setPasswordHash(createHash(newPassword));
+        return true;
+    }
+
+    @Override
+    public Trainer createTrainer(Trainer trainer, String password) throws NoAdministratorException {
+        //if adding trainer that is not an admin check if there is at least one already
+        if (!trainer.isAdmin() && trainerDao.getAdminCount() == 0) {
```

Comment:
I think there should be always at least one admin, so I would remove the first part of the is (before &&).

Reply:
The validation should be executed only if the admin flag is being set to false.
If I remove the condition and there is just one trainer and that is admin and I try to set admin true again an Exception would be thrown and that would be incorrect.